docs/design/security_context.md

```diff
@@ -0,0 +1,158 @@
+    * Each namespace is associated with a range of uids/gids on the node that
+      are mapped to uids/gids on containers using linux user namespaces. 
+    * Certain pods in each namespace have special privileges to perform system
+      actions such as talking back to the server for deployment, run docker
+      builds, etc.
+    * External NFS storage is assigned to each namespace and permissions set
+      using the range of uids/gids assigned to that namespace. 
+
+## Proposed Design
+
+### Overview
+A *security context* consists of a set of constraints that determine how a container
+is secured before getting created and run. It has a 1:1 correspondence to a
+[service account](https://github.com/GoogleCloudPlatform/kubernetes/pull/2297). A *security context provider* is passed to the Kubelet so it can have a chance
+to mutate Docker API calls in order to apply the security context.
+
+It is recommended that this design be implemented in two phases:
+
+1.  Implement the security context provider extension point in the Kubelet 
+    so that a default security context can be applied on container run and creation.
+2.  Implement a security context structure that is part of a service account. The
+    default context provider can then be used to apply a security context based
+    on the service account associated with the pod.
+    
+### Security Context Provider
+
+The Kubelet will have an interface that points to a `SecurityContextProvider`. The `SecurityContextProvider` is invoked before creating and running a given container:
+
+```go
+type SecurityContextProvider interface {
+    // ModifyContainerConfig is called before the Docker createContainer call.
+    // The security context provider can make changes to the Config with which
+    // the container is created.
+    // An error is returned if it's not possible to secure the container as 
+    // requested with a security context. 
+	ModifyContainerConfig(pod *api.BoundPod, container *api.Container, config *docker.Config) error
+	
+	// ModifyHostConfig is called before the Docker runContainer call.
+	// The security context provider can make changes to the HostConfig, affecting
+	// security options, whether the container is privileged, volume binds, etc.
+	// An error is returned if it's not possible to secure the container as requested 
+    // with a security context. 
+	ModifyHostConfig(pod *api.BoundPod, container *api.Container, hostConfig *docker.HostConfig)
+}
+```
+If the value of the SecurityContextProvider field on the Kubelet is nil, the kubelet will create and run the container as it does today.   
+
+### Security Context
+
+A security context has a 1:1 correspondence to a service account and it can be included as
+part of the service account resource. Following is an example of an initial implementation:
+
+```go
+type SecurityContext struct {
+    // user is the uid to use when running the container
+	User int
```

Comment:
----- Original Message -----
> > +	// An error is returned if it's not possible to secure the container as
> > requested
> > +    // with a security context.
> > +	ModifyHostConfig(pod *api.BoundPod, container *api.Container, hostConfig
> > *docker.HostConfig)
> > +}
> > +```
> > +If the value of the SecurityContextProvider field on the Kubelet is nil,
> > the kubelet will create and run the container as it does today.
> > +
> > +### Security Context
> > +
> > +A security context has a 1:1 correspondence to a service account and it
> > can be included as
> > +part of the service account resource. Following is an example of an
> > initial implementation:
> > +
> > +```go
> > +type SecurityContext struct {
> > +    // user is the uid to use when running the container
> > +	User int
> 
> One option is to declare that we don't support this style of container, but
> that excludes a lot of existing images, I think.

Some, but that could also be the old school image (pre-pods) vs the new school (one process / user per container).  We can probably get pretty far on that for Kube users.

> 
> Another option is to just set the lead process to have this uid, but allow
> other processes to have other virtual uids which map back to useless
> physical uids.
> 
> Another option, which I'm not sure if it works at all, is to map multiple
> virtual uids to 1 physical uids.

I don't think it works.  In docker upstream we had a long discussion on this about ranges - we *think* we can allocate ranges and have this work, but you have to have large ranges.  If people had to predeclare how many uids they need and had a map or something, we could maybe allocate a set for a namespace (10k was mooted per container before).  I've also wondered whether we could just do two ranges - shared, and unshared.  Shared is allocated by the master and cluster wide.  Unshared is node scoped and each started container gets a set.  I *think* you can then pass two ranges into the container.  @mrunalp to keep me honest here.

> 
> (I'm using virtual to mean "in-namespace" and physical to mean "in the root
> linux namespace").
> 
> Another option is to use a per-volume strategy for virtual-to-physical
> mapping.
> For NFS, you could use the NFS mount options to map a single local uid to a
> true remote uid using anonuid and anongid.  Then you would map all the
> container uids to the local anongid.  Like this:
> `
> container uid 0 maps to rootns uid 12345 using usernamespaces.  rootns uid
> 12345 maps to NFS server uid 5432 using mount options.
> `
> Note that this does require an nfs mount for every container versus one mount
> per node, but I think that is the way y'all were planning anyhow.
> 
> 
> I think higher level questions is: If there are two uids in a container,
> should their filesystem writes, at the canonical view of the filesystem,
> appear as one or two different uids/gids?  I think "one" is simpler for
> users but harder for implementers.

Yeah, although in practice I suspect 60-80% of containers that people *should* run will be single uid.  So we can make single uid work well, and have multi uid be not quite as nice.  

> 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/3910/files#r23854292